src/compiler/checker.ts

```diff
@@ -28251,14 +28251,17 @@ namespace ts {
             // of the property as base classes
             let enclosingClass = forEachEnclosingClass(location, enclosingDeclaration => {
                 const enclosingClass = getDeclaredTypeOfSymbol(getSymbolOfNode(enclosingDeclaration)!) as InterfaceType;
-                return isClassDerivedFromDeclaringClasses(enclosingClass, prop, writing) ? enclosingClass : undefined;
+                return isClassDerivedFromDeclaringClasses(enclosingClass, prop, writing);
             });
             // A protected property is accessible if the property is within the declaring class or classes derived from it
             if (!enclosingClass) {
                 // allow PropertyAccessibility if context is in function with this parameter
-                // static member access is disallow
-                let thisParameter: ParameterDeclaration | undefined;
-                if (flags & ModifierFlags.Static || !(thisParameter = getThisParameterFromNodeContext(location)) || !thisParameter.type) {
+                // static member access is disallowed
+                if (
+                    flags & ModifierFlags.Static
+                    || !(enclosingClass = getEnclosingClassFromThisParameter(location))
+                    || !(enclosingClass = isClassDerivedFromDeclaringClasses(enclosingClass, prop, writing))
+                ) {

```

Comment:
If it's OK to run the checks even for static methods (and I think it is), then the below is much less surprising to read:

```suggestion
                enclosingClass = getEnclosingClassFromThisParameter(location)) || isClassDerivedFromDeclaringClasses(enclosingClass, prop, writing);
                if (flags & ModifierFlags.Static || !enclosingClass) {
```

Reply:
Yeah should be fine, at best it was just a very minor optimization to avoid fetching `this`'s parameter if the field was already static.

Reply:
`getEnclosingClassFromThisParameter` can return `undefined` while `isClassDerivedFromDeclaringClasses` doesn't accept `undefined`, so I'll have to do some type checking with this style instead:
```typescript
enclosingClass = getEnclosingClassFromThisParameter(location);
enclosingClass = enclosingClass && isClassDerivedFromDeclaringClasses(enclosingClass, prop, writing);
if (flags & ModifierFlags.Static || !enclosingClass) {
```

Reply:
Also noticed that the name `isClassDerivedFromDeclaringClasses` may be a bit misleading, as it doesn't return a `boolean`, but instead it returns the first parameter if checks succeeds, otherwise it returns undefined.

Not sure if we'd prefer changing the name or not.

Reply:
Maybe as a follow-up, I'd rather get this change in.